Bluetooth: controller: Fix missing storage for sync cancel reason

cvinayak opened the pull request:
Fix missing storage allocation in ll_sync_set structure for
storing the sync cancel reason.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix missing storage for sync cancel reason

Fix missing storage allocation in ll_sync_set structure for
storing the sync cancel reason.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -205,17 +205,17 @@ uint8_t ll_sync_create_cancel(void **rx)
 	ll_rx_release(node_rx);
 
-	node_rx = (void *)&sync->node_rx_lost;
+	node_rx = (void *)&sync->node_rx_lost.hdr;
```
in `ull_conn.c:1716` the node_rx is assigned to the outer struct, rather than the header. 

I think we should consider being consistent with that, so either do the same here, or change the `ull_conn.c` as well.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -205,17 +205,17 @@ uint8_t ll_sync_create_cancel(void **rx)
 	node_rx->hdr.handle = 0xffff;
 	node_rx->hdr.rx_ftr.param = sync;
-	se = (void *)node_rx->pdu;
+	se = (void *)&sync->node_rx_lost.reason;
```
I'm not sure I understand this. 

`se` is of type `struct node_rx_sync` and `sync->node_rx_lost.reason` is just a `uint8_t`. By doing this, you make `status == reason`, but if trying to access any other member of the `se` struct, you access illegal data.

cvinayak committed:
Bluetooth: controller: Make conn cancel and sync cancel impl. consistent

Make the connection cancel and periodic advertising sync
create cancel implementation consistent.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@wopu-ot request your review. (do you think we should add you into code owners file, so that you are added automatically?)

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -405,10 +404,17 @@ uint8_t ll_connect_disable(void **rx)
 		cc->hdr.type = NODE_RX_TYPE_CONNECTION;
 		cc->hdr.handle = 0xffff;
+
+		/* NOTE: struct llcp_terminate.node_rx has uint8_t member
+		 *       following the struct node_rx_hdr to store the reason.
+		 */
 		*((uint8_t *)cc->pdu) = BT_HCI_ERR_UNKNOWN_CONN_ID;
```
That's quite ugly and should be cleaned up. However, that should happen in a different PR (documenting ugliness should not be "rewarded" with scope creep).

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -203,22 +202,29 @@ uint8_t ll_sync_create_cancel(void **rx)
-	node_rx->hdr.rx_ftr.param = sync;
-	se = (void *)node_rx->pdu;
-	se->status = BT_HCI_ERR_OP_CANCELLED_BY_HOST;
+	/* NOTE: struct node_rx_lost has uint8_t member following the
+	 *       struct node_rx_hdr to store the reason.
+	 */
+	*((uint8_t *)se->pdu) = BT_HCI_ERR_OP_CANCELLED_BY_HOST;
```
This is quite ugly. I would prefer to have the type of the `node_rx_lost` field be non-anonymous so `se` can be declared with that type and the `reason` field is accessible. I would also be okay with accessing `sync->node_rx_lost.reason` directly.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -173,10 +173,9 @@ uint8_t ll_sync_create_cancel(void **rx)
 	struct ll_scan_set *scan_coded;
 	memq_link_t *link_sync_estab;
 	memq_link_t *link_sync_lost;
-	struct node_rx_pdu *node_rx;
```
The compiler is likely smart enough that removing this auto variable does not save anything but only makes the code a little harder to read.

cvinayak committed:
Bluetooth: controller: Use connection complete variable naming

Use the connection complete varaible naming to access
connection complete parameters.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

wopu-ot reviewed (commented):
After the changes and re-reviewing, I think I understand a little more. Do I understand correctly that the `reason` field in `node_rx_lost` must be at the same offset as the `status` field in a `node_rx_pdu` with a `node_rx_sync` payload? A comment regarding this at the places where `node_rx_lost` and `node_rx_sync` are declared would be helpful to avoid future changes breaking this assumption.

Additionally, I realized that the `pdu` field in `node_rx_pdu` may be unaligned but `node_rx_sync` may have alignment requirements. Assignments like `struct node_rx_sync *se = (void *)node_rx->pdu` are therefore bound to break at some point. This does not necessarily need to be resolved in this PR, but should be resolved eventually.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -214,10 +214,18 @@ uint8_t ll_sync_create_cancel(void **rx)
 	node_rx = (void *)&sync->node_rx_lost;
 	node_rx->hdr.type = NODE_RX_TYPE_SYNC;
 	node_rx->hdr.handle = 0xffff;
-	node_rx->hdr.rx_ftr.param = sync;
+
+	/* NOTE: struct node_rx_lost has uint8_t member following the
+	 *       struct node_rx_hdr to store the reason.
+	 */
 	se = (void *)node_rx->pdu;
 	se->status = BT_HCI_ERR_OP_CANCELLED_BY_HOST;
 
+	/* NOTE: Since NODE_RX_TYPE_SYNC is only generated from ULL context,
+	 *       pass ULL context as parameter.
+	 */
+	node_rx->hdr.rx_ftr.param = sync;
+
 	*rx = node_rx;
```
I guess this is equivalent to writing
```
sync->node_rx_lost.hdr.type = NODE_RX_TYPE_SYNC;
sync->node_rx_lost.hdr.handle = 0xffff;
sync->node_rx_lost.hdr.rx_ftr.param = sync;
sync->node_rx_lost.reason = BT_HCI_ERR_OP_CANCELLED_BY_HOST;

*rx = &sync->node_rx_lost;
```
Isn't it? 

I wonder if this is more readable, as you're directly changing the values of the fields in the `sync->node_rx_lost`, rather than doing it through 2 other pointers. 

Similar comment also applies for the `ll_connect_disable` change above.

cvinayak replied:
The code is kept similar to HCI access to the reason, this way future code/features can follow the design.

Thalley replied:
I understand that, and will not block the PR based on that. 

However, if using the fields of the `sync->node_rx_lost` it would be possible, if ever needed, to expand that easier than if just expected a certain sequence of fields in that struct. But it's alright as it is right now.

cvinayak commented:
>A comment regarding this at the places where `node_rx_lost` and `node_rx_sync` are declared would be helpful to avoid future changes breaking this assumption.

Will do.

> Additionally, I realized that the `pdu` field in `node_rx_pdu` may be unaligned but `node_rx_sync` may have alignment requirements. Assignments like `struct node_rx_sync *se = (void *)node_rx->pdu` are therefore bound to break at some point. 

Will address this in a separate PR.

cvinayak committed:
Bluetooth: controller: Fix memory alignment of PDU buffer reference

Align the PDU buffer reference in struct node_rx_pdu so that
node rx type specific parameters, like, terminate and sync
lost reason can be accessed without any memory alignment
issues.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -281,9 +281,16 @@ struct node_rx_hdr {
 	struct node_rx_hdr hdr;
-	uint8_t               pdu[0];
+	union {
+		uint8_t    pdu[0] __aligned(4);
```
I'm not sure I understand why the `__aligned(4)` part is necessary. Wouldn't the the `pdu[0]` fields be aligned correctly without it?

cvinayak replied:
`pdu[0]` is uint8_t and hence depending on the previous structure member size, would be placed unaligned (packed). HCI code typecasts unpacked structures to `&pdu[0]`, potentially leading to unaligned access faults if `struct node_rx_hdr` size changes.

cvinayak replied:
Alternatively, the below too solves the problem.

```suggestion
	union {
		uint8_t    pdu[0];
		uint32_t   param[0];
	};
```

Thalley replied:
I see. Wouldn't defining pdu as a `void *` provide the same? Do we lose anything from doing that? 

It's not an issue that it's there, but I guess we'll be casting `pdu` to another struct definition anyhow, so does it matter that it's uint8_t or void?

wopu-ot replied:
Defining pdu as void* would always occupy 4 bytes, even when not needed.

I think having the zero-length array with `__aligned(4)` is the correct solution. I am not sure whether the union in the `node_rx_pdu` struct is strictly necessary, but it is consistent with how the other struct look like, which is a good thing.

Thalley replied:
You are correct :) It could be `uint32_t pdu[0]` though, but I think that `__aligned(4)` serves the same purpose just as well.
